### DISCUSSION TITLE
Minimal API questions related to streaming response

### DISCUSSION THREAD

**Discussion by cirrusone** [dotnet/aspnetcore#54320] [Q&A] [answered]
Two questions...

1/ Does minimal API automatically dispose of `Stream` similar to MVC or do I need to handle disposal of Stream? I.e. is the following okay?

```
    static void Main(string[] args)
    {
        MessagePackSerializerOptions lz4Options = MessagePackSerializerOptions.Standard.WithCompression(MessagePackCompression.Lz4BlockArray);
        var builder = WebApplication.CreateBuilder(args);
        var app = builder.Build();

        app.MapPost("/calculate", async (Stream body) =>
        {
            Data data = await MessagePackSerializer.DeserializeAsync<Data>(body, lz4Options);
            // Does the following Stream need to be disposed?
            Stream stream = new MemoryStream(MessagePackSerializer.Serialize(new DataResult() { Result = data.X * data.Y }, lz4Options));
            return Results.Stream(stream);
        });

        app.Run();
    }
```

2/ In MVC I use the following

```
    [HttpPost("[action]")]
    [Produces("application/octet-stream")]
    public async Task Calculate()
    {
        MessagePackSerializerOptions lz4Options = MessagePackSerializerOptions.Standard.WithCompression(MessagePackCompression.Lz4BlockArray);
        Data data = await MessagePackSerializer.DeserializeAsync<Data>(body, lz4Options);
        await MessagePackSerializer.SerializeAsync(Response.BodyWriter.AsStream(false), new DataResult() { Result = data.X * data.Y }, _lz4Options);
    }
```

What's the exact representation of this in minimal API or have I managed to replicate this exactly in 1/ above? Minimal API does not have `Response.BodyWriter`.

**Comment by davidfowl** [comment, marked as answer]
There’s an overload of stream that gives you a callback Func\<Stream, Task\>

**Reply by cirrusone** [reply]
Ah excellent, that's really helpful. Is this quite new as I couldn't find this in the docs? I may have been searching in wrong place perhaps.

This works well for me 
`return Results.Stream(stream => MessagePackSerializer.SerializeAsync(stream, new DataResult() { Result = data.X * data.Y }, lz4Options));`

**Reply by davidfowl** [reply]
Added in .NET 7 https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.results.stream?view=aspnetcore-7.0
